Destructuring Assignment Design and Milestone 1 code

techno96 opened the pull request:
Members : 

- Divija Devarla
- Purva Kothari
- Subha Ramesh

GV1028 reviewed (commented):
Good work!  Here are a few suggestions:

1. I feel that the design choices would not scale up to future extensions. Consider the following example:
```
def f(n : int) -> int:
  a = range(0, n)
  v1, v2, v3, v4, v5 = a
  return v1 + v2 + v3 + v4 + v5
f(5)
f(3) # throws a dynamic error "too many values to unpack"
```
These cases would also fail type checking for length of rhs and lhs.

2. Also what would be the right way to approach  _  and * in destructuring assignments? Do you need to make changes to AST for these?

```
a, *b, c = range(0,4)
assert (a == 0)
assert( b == [1,2])
assert( c == 3)
```

techno96 commented:
@GV1028 , Thank you for the valuable comments. We have addressed the comments and updated our design to include these as well. W.r.t our AST changes, we have 2 ideas which we have documented. We would like your opinion on which idea will be extensible down the line.

GV1028 commented on ast.ts:
```diff
@@ -56,3 +58,9 @@ export enum UniOp { Neg, Not };
+
+export type DestructureLHS<A> = { a?: A, lhs: AssignTarget<A>, isStarred : boolean, isIgnore : boolean}
+
+export type AssignTarget<A> = 
```
Just making sure I understand, can this design choice in AST accommodate a,b,c = [1,2,3]?

pk38 replied:
Thank you for all your valuable input!
The current design will support destructuring a,b,c but since lists are not implemented, it will currently support a,b,c = 1,2,3.

We made a conscious decision to stick to the above cases in order to reduce duplicated efforts. As part of the merging changes, we will extend support to lists, sets, dicts on RHS. The current changes for LHS will by default support them.

We had added all the Milestone 1 working examples in a new design.md file - https://github.com/ucsd-cse231-s22/chocopy-wasm-compiler-B/pull/8/files#diff-7e6258a05da56059e61588bda0ee015fc5486f0f0e9c757c080f0eb005cbd0ba

All the examples listed here are currently working.

GV1028 commented on lower.ts:
```diff
@@ -99,7 +99,68 @@ function flattenStmt(s : AST.Stmt<[Type, SourceLocation]>, blocks: Array<IR.Basi
+      // }
 
+      // return rhsinits
     case "return":
```
Clean up all stray comments before PR to avoid unnecessary merge conflicts

techno96 replied:
We have cleaned up all the stray comments.

GV1028 commented on ast.ts:
```diff
@@ -42,6 +43,7 @@ export type Expr<A> =
   | {  a?: A, tag: "index", obj: Expr<A>, index: Expr<A> }
   | {  a?: A, tag: "method-call", obj: Expr<A>, method: string, arguments: Array<Expr<A>> }
   | {  a?: A, tag: "construct", name: string }
+  | {  a?: A, tag: "tuple", expr: Expr<A>[]}
```
I see you went ahead with a tuple implementation. I am not sure if this is part of your previous milestone goals. I would suggest interacting with the sets/tuples team to avoid any conflicts. I am pretty sure there will be a merge conflict.

If you need an iterator to work with, working with a range class would be the best option. Any reason why you didn't go that route?

techno96 replied:
We did not support Range in this milestone in order to re-use the For loops implementation of Range class. We intend to support that as part of the final Merge changes on Monday, by pulling in their changes. For Milestone 1, we have implemented all the basic cases (including ignore and starred) outside of Iterators and Data Structures.

We have thus removed instances of Range and Tuple (supported by other teams) from our commits to avoid conflicts.

techno96 replied:
We have now added support for Range in this pull request.

GV1028 commented on parser.ts:
```diff
@@ -197,6 +197,21 @@ export function traverseExpr(c : TreeCursor, s : string) : Expr<SourceLocation>
+    case "TupleExpression":
+      let tuple :Expr<SourceLocation>[] = [];
+      c.firstChild(); // Open parenthesis "("
+      c.nextSibling();
```
Again merge conflict warning, please talk to the sets/tuples team of compiler B

techno96 replied:
This has been removed.

GV1028 commented on parser.ts:
```diff
@@ -348,6 +388,57 @@ export function traverseStmt(c : TreeCursor, s : string) : Stmt<SourceLocation>
+  const lhs = traverseExpr(c,s);
+  // 2. LHS is valid expression type : "id" | "lookup" 
+  if (lhs.tag!=="id" && lhs.tag!=="lookup") {
+    throw new ParseError("Cannot have "+ lhs.tag + " expression at LHS while parsing assignment statements.", location.line)
```
Hmm seems like you are doing a bit of error reporting as well, again possibilities of conflicts. Please do talk to the error reporting team as well.

techno96 replied:
We are working with the error reporting team to iron this out.

GV1028 commented on type-check.ts:
```diff
@@ -25,10 +25,22 @@ defaultGlobalFunctions.set("min", [[NUM, NUM], NUM]);
+rangeMethods.set("__init__", [{tag: "number"}, {tag: "number"}, {tag: "class"}])
+rangeMethods.set("hasNext", [{tag: "class"}])
+rangeMethods.set("next", [{tag: "class"}])
+defaultGlobalClasses.set("range", [rangeFields, rangeMethods]);
```
Is this used anywhere? I am not sure I understand clearly if you are working on tuples or the range class.

techno96 replied:
This has been removed now. Please refer the earlier comment detailing our decision : https://github.com/ucsd-cse231-s22/chocopy-wasm-compiler-B/pull/8#discussion_r877663827

GV1028 reviewed (changes requested):
Good job! But I would like to see some explanations on some of the design choices to better understand if you are using the range class. I am not sure I completely understand.

I think there will also be a few merge conflicts with the sets/tuples group because you modify the AST to include your idea of tuples. Please do talk with the group to see if they are working on something that you could use. But for now, I would just work with something like ``range`` so that you have an iterator to work with. I am also not sure where exactly the ``range`` class is being used. In type checking, you want to avoid statically checking the lengths of RHS and LHS and instead use the ``range`` methods to generate code that will check lengths and do the appropriate iteration and assignment at runtime.

KamranAlipour commented on type-check.ts:
```diff
@@ -223,6 +238,62 @@ export function tcStmt(env : GlobalTypeEnv, locals : LocalTypeEnv, stmt : Stmt<S
 
+export function tcDestructure(env : GlobalTypeEnv, locals : LocalTypeEnv, destr : DestructureLHS<SourceLocation>) : DestructureLHS<[Type, SourceLocation]> {
+  
+  if (destr.lhs.tag === "id" && destr.lhs.name === "_") {
```
Can you please leave a comment and explain what is done here?

techno96 replied:
The comment has been [added](https://github.com/ucsd-cse231-s22/chocopy-wasm-compiler-B/pull/8/files#diff-0a03798f0e47b1080516ea97c55cfc6c2e025276ef50bf1e1ede9b6e4570a33cR261).

KamranAlipour commented on tests/destructuring.test.ts:
```diff
@@ -0,0 +1,164 @@
+print(y)
+print(z)`)
+
+});
```
Good tests, I think you might also wanna include a couple of tests that destructuring happens within a function definition, if it uses some function parameters it would be even better!

1Divija1 replied:
Thank you for the feedback. We added a basic testcase using function Params. We will be adding more as part of milestone 2.

techno96 commented:
Closing this pull request as the newer ones have been merged.
